### DISCUSSION TITLE
es/integrations/onnx/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#14654] [Docs]
# es/integrations/onnx/

Más información sobre cómo exportar YOLOv8 modelos para ONNX para una implementación flexible en varias plataformas con un rendimiento mejorado.

https://docs.ultralytics.com/es/integrations/onnx/

**Comment by wisrovi** [comment]
tengo el siguiente codigo:

``` python
from ultralytics import RTDETR
from ultralytics import YOLO


DETECTOR_PATH = "./src/models/detector/"
DETECTOR_MODEL = "my_model"  # este modelo esta entrenado con imgz1024 usando RTDETR-X como detector


model = RTDETR(f"{DETECTOR_PATH}{DETECTOR_MODEL}.pt")


"""


            haciendo una primera inferencia del modelo


"""


results_original = model("data/images/009.jpg", verbose=False)


"""


    convirtiendo el modelo a onnx


"""
model.export(
    format="onnx",
    # int8=True,
    verbose=True,
    # imgsz=1024,
    dynamic=True,
    batch=32,
    data="data.yaml",
)
"""
Esto imprime:

Ultralytics YOLOv8.2.49 🚀 Python-3.12.1 torch-2.2.1+cu121 CPU (Intel Core(TM) i7-9700F 3.00GHz)
rt-detr-x summary: 642 layers, 65469491 parameters, 0 gradients, 222.5 GFLOPs

PyTorch: starting from 'src/models/detector/my_model.pt' with input shape (32, 3, 1024, 1024) BCHW and output shape(s) (32, 300, 5) (129.1 MB)

ONNX: starting export with onnx 1.16.1 opset 17...
ONNX: export success ✅ 248.9s, saved as 'src/models/detector/my_model.onnx' (250.3 MB)

Export complete (723.3s)
Results saved to ./src/models/detector/
Predict:         yolo predict task=detect model=src/models/detector/my_model.onnx.onnx imgsz=1024  
Validate:        yolo val task=detect model=src/models/detector/my_model.onnx.onnx imgsz=1024 data=/dataset/data.yaml  
Visualize:       https://netron.app
"""



"""


        probando que la conversion me da los mismos resultados


"""


model = YOLO(f"{DETECTOR_PATH}{DETECTOR_MODEL}.onnx")
results_yolo = model("data/images/009.jpg", verbose=False)

model = RTDETR(f"{DETECTOR_PATH}{DETECTOR_MODEL}.onnx")
results_rtdetr = model("data/images/009.jpg", verbose=False)

"""
Esto imprime:

WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Loading src/models/detector/my_model.onnx for ONNX Runtime inference...
2024-07-24 11:47:45.285125009 [E:onnxruntime:Default, provider_bridge_ort.cc:1745 TryGetProviderInfo_CUDA] /onnxruntime_src/onnxruntime/core/session/provider_bridge_ort.cc:1426 onnxruntime::Provider& onnxruntime::ProviderLibrary::Get() [ONNXRuntimeError] : 1 : FAIL : Failed to load library libonnxruntime_providers_cuda.so with error: libcublasLt.so.11: cannot open shared object file: No such file or directory

2024-07-24 11:47:45.285143619 [W:onnxruntime:Default, onnxruntime_pybind_state.cc:895 CreateExecutionProviderInstance] Failed to create CUDAExecutionProvider. Please reference https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirementsto ensure all dependencies are met.
Loading src/models/detectormy_model.onnx for ONNX Runtime inference...
2024-07-24 11:47:51.875097838 [E:onnxruntime:Default, provider_bridge_ort.cc:1745 TryGetProviderInfo_CUDA] /onnxruntime_src/onnxruntime/core/session/provider_bridge_ort.cc:1426 onnxruntime::Provider& onnxruntime::ProviderLibrary::Get() [ONNXRuntimeError] : 1 : FAIL : Failed to load library libonnxruntime_providers_cuda.so with error: libcublasLt.so.11: cannot open shared object file: No such file or directory

2024-07-24 11:47:51.875114713 [W:onnxruntime:Default, onnxruntime_pybind_state.cc:895 CreateExecutionProviderInstance] Failed to create CUDAExecutionProvider. Please reference https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirementsto ensure all dependencies are met.
"""



print("Original")
for result in results_original:
    for confidence, cls, box in zip(
        result.boxes.conf, result.boxes.cls, result.boxes.xyxy
    ):
        box = [
            int(box[0]),
            int(box[1]),
            int(box[2]),
            int(box[3]),
        ]

        print(int(confidence), box, int(cls))


print("Converted and yolo")
for result in results_yolo:
    for confidence, cls, box in zip(
        result.boxes.conf, result.boxes.cls, result.boxes.xyxy
    ):
        box = [
            int(box[0]),
            int(box[1]),
            int(box[2]),
            int(box[3]),
        ]

        print(int(confidence), box, int(cls))


print("Converted and rtdetr")
for result in results_rtdetr:
    for confidence, cls, box in zip(
        result.boxes.conf, result.boxes.cls, result.boxes.xyxy
    ):
        box = [
            int(box[0]),
            int(box[1]),
            int(box[2]),
            int(box[3]),
        ]

        print(int(confidence), box, int(cls))

```


la conversion del modelo se realiza y se crea el model.onnx, pero los resultados son diferentes en las 3 detecciones:

```python
Original
0 [48, 518, 149, 611] 0
0 [33, 575, 59, 616] 0
0 [1633, 526, 1722, 606] 0
Converted and yolo
0 [0, 0, 0, 0] 108
0 [0, 0, 1, 0] 206
0 [0, 0, 0, 0] 185
Converted and rtdetr

```

necesito realizar la conversion para luego montar el modelo.onnx con Triton Servidor de inferencia con Ultralytics YOLOv8 o con Neural Magic's DeepSparse Engine

Aún no decido cual es mejor para mi solución, pero dadas las predicciones distintas despues de importar el modelo, esto me preocupa de haber hecho mal la conversion.

Podrian ayudarme por favor

**Reply by glenn-jocher** [reply]
@wisrovi gracias por compartir tu código y detalles del problema. Parece que el error está relacionado con la falta de la biblioteca `libonnxruntime_providers_cuda.so` en tu entorno, lo que impide la creación del `CUDAExecutionProvider`. Para resolver esto, asegúrate de que todas las dependencias de ONNX Runtime para CUDA estén instaladas correctamente. Puedes verificar los requisitos y la instalación en la [documentación de ONNX Runtime](https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirements).

Además, asegúrate de estar utilizando la última versión de Ultralytics y ONNX Runtime. Si el problema persiste después de verificar las dependencias y actualizar las versiones, por favor háznoslo saber para seguir investigando.

**Reply by wisrovi** [reply]
he vuelto a probar, esta vez intente con un modelo de yolov8n,pt, con el mismo código, y este ha funcionado bien, pero cuando intento con mi modelo pesado de RTDETR-x, este no funciona

hay alguna restriccion de tamaño al convertir a onnx?

o quizas, al ser un trasnformers la conversion no se puede realizar?

me ayudan, por favor?

**Reply by glenn-jocher** [reply]
Gracias por tu comentario. No hay restricciones de tamaño específicas al convertir modelos a ONNX, pero los modelos más grandes y complejos, como RTDETR-x, pueden enfrentar problemas debido a la complejidad de la arquitectura y la cantidad de parámetros. Asegúrate de estar utilizando la última versión de Ultralytics y ONNX Runtime. También, verifica que todas las dependencias de ONNX Runtime estén correctamente instaladas, especialmente si estás utilizando CUDA. Si el problema persiste, intenta simplificar el modelo o ajustar los parámetros de exportación. Para más detalles, visita nuestra [documentación sobre exportación a ONNX](https://docs.ultralytics.com/integrations/onnx/).

**Reply by wisrovi** [reply]
Para crear mi entorno uso:

```python
sudo docker run -it --rm --gpus all -e DISPLAY=unix$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix --shm-size=16g -v ./:/dataset ultralytics/ultralytics:8.2.45 bash
```

y adentro del entorno, actualizo para tener las ultimas versiones de la libreria
```python
pip install --upgrade ultralytics
```

por supuesto que antes he instalado los drivers de nvidia-docker para poder compartir la GPU, y lo se porque cuando uso yolov8n.pt, todo funciona bien, el problema es cuando convierto el  RTDETR-x (con 65M de parametros), he hecho experimentos con:
```python
model.export(
    format="onnx",
    # int8=True,
    verbose=True,
    # imgsz=1024,
    dynamic=True,
    batch=32,
    data="data.yaml",
    opset=16 
)
```
cambiando los valores de los parametros de la exportación, el opset=16 es el único en donde la conversión puede llevarse a cabo, pero no se que otros parametros podria editar para lograr mi cometido

Estoy haciendo algo mal? o me falta algo por hacer?

**Reply by wisrovi** [reply]
@glenn-jocher

**Reply by glenn-jocher** [reply]
It seems like the issue might be related to the complexity of the RTDETR-x model. Ensure your Docker environment has sufficient resources and try reducing the batch size or using a different opset version. If the problem persists, consider simplifying the model architecture. For more details, visit our ONNX export documentation.
